analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java

```diff
@@ -0,0 +1,1261 @@
+            if (list == null) {
+                refresh();
+                Display.getDefault().asyncExec(() -> buildFlameGraph(trace, null, null));
+
+            } else {
+                // Reset end time
+                long endTime = Long.MIN_VALUE;
+                for (TimeGraphEntry entry : list) {
+                    endTime = Math.max(endTime, entry.getEndTime());
+                }
+                setEndTime(endTime);
+                refresh();
+                startZoomThread(0, endTime, false);
+            }
+        }
+    }
+
+    /**
+     * Get the callgraph modules used to build the view
+     *
+     * @return The call graph provider modules
+     */
+    protected Iterable<ICallGraphProvider2> getCallgraphModules() {
+        ITmfTrace trace = fTrace;
+        if (trace == null) {
+            return Collections.emptyList();
+        }
+        String analysisId = NonNullUtils.nullToEmptyString(getViewSite().getSecondaryId());
+        Iterable<ICallGraphProvider2> modules = TmfTraceUtils.getAnalysisModulesOfClass(trace, ICallGraphProvider2.class);
+        return StreamSupport.stream(modules.spliterator(), false)
+                .filter(m -> {
+                    if (m instanceof IAnalysisModule analysisModule) {
+                        return analysisModule.getId().equals(analysisId);
+                    }
+                    return true;
+                })
+                .collect(Collectors.toSet());
+    }
+
+    private class BuildRunnable {
+        private final ITmfTrace fBuildTrace;
+        private final ITmfTrace fParentTrace;
+        private final FlowScopeLog fScope;
+        private final Map<String, Object> fParameters;
+
+        public BuildRunnable(final ITmfTrace trace, final ITmfTrace parentTrace, @Nullable ITmfTimestamp selStart, @Nullable ITmfTimestamp selEnd, final FlowScopeLog log) {
+            fBuildTrace = trace;
+            fParentTrace = parentTrace;
+            fScope = log;
+            if (selStart != null && selEnd != null) {
+                fParameters = ImmutableMap.of(FlameGraphDataProvider.SELECTION_RANGE_KEY, ImmutableList.of(selStart.toNanos(), selEnd.toNanos()));
+            } else {
+                fParameters = Collections.emptyMap();
+            }
+        }
+
+        public void run(IProgressMonitor monitor) {
+            try (FlowScopeLog log = new FlowScopeLogBuilder(LOGGER, Level.FINE, "FlameGraphView:BuildThread", "trace", fBuildTrace.getName()).setParentScope(fScope).build()) { //$NON-NLS-1$ //$NON-NLS-2$
+                buildEntryList(fBuildTrace, fParentTrace, fParameters, Objects.requireNonNull(monitor));
+                synchronized (fBuildJobMap) {
+                    fBuildJobMap.remove(fBuildTrace);
+                }
+            }
+        }
+
+        private void buildEntryList(@Nullable ITmfTrace trace, ITmfTrace parentTrace, Map<String, Object> additionalParams, IProgressMonitor monitor) {
```

Comment:
Refactor this method to reduce complexity, lines of code and nesting level.

Reply:
Done, its much better than the initial implementation, but we are still at a Cognitive Complexity level of 18. 
I am struggling to refactor it more than that, what do you think ?